Replace illegal characters

Feature request: Replace illegal characters (e.g., replace a pipe with a hyphen) rather than forcing a manual rename. Most of the other Commons importers do this. Thanks for your continued support of this great tool

(Also a small typo in the error message: occured → occurred)

Example error:
`Error: An exception occured: AssertionError: Your filename contains an illegal character: |`

--- a/video2commons/frontend/api.py
+++ b/video2commons/frontend/api.py
@@ -297,7 +297,7 @@ def validate_filename():
 def _validate_filename(filename):
     illegalords = range(0, 32) + [127]
     for char in filename:
-        assert char not in '[]{}|#<>%+?!:/\\.', \
+        assert char not in '[]{}|#<>%+?:/\\', \
             'Your filename contains an illegal character: ' + char
 
         # ord(char) to prevent bad renderings

--- a/video2commons/frontend/api.py
+++ b/video2commons/frontend/api.py
@@ -34,7 +34,7 @@ from video2commons.exceptions import NeedServerSideUpload
 
 from video2commons.frontend.shared import redisconnection, check_banned
 from video2commons.frontend.urlextract import (
-    do_extract_url, do_validate_filename
+    do_extract_url, do_validate_filename, sanitize
 )
 
 api = Blueprint('api', __name__)
@@ -323,7 +323,7 @@ def run_task():
     url = request.form['url']
     ie_key = request.form['extractor']
     subtitles = request.form['subtitles']
-    filename = request.form['filename']
+    filename = sanitize(request.form['filename'])
     filedesc = request.form['filedesc']
     downloadkey, convertkey = get_backend_keys(request.form['format'])
     username = session['username']

--- a/video2commons/frontend/api.py
+++ b/video2commons/frontend/api.py
@@ -21,7 +21,6 @@
 
 from __future__ import absolute_import
 
-import re
 import pickle
 import traceback
 import urllib
@@ -34,7 +33,9 @@ from video2commons.backend import worker
 from video2commons.exceptions import NeedServerSideUpload
 
 from video2commons.frontend.shared import redisconnection, check_banned
-from video2commons.frontend.urlextract import do_extract_url
+from video2commons.frontend.urlextract import (
+    do_extract_url, do_validate_filename
+)
 
 api = Blueprint('api', __name__)
 
@@ -290,26 +291,8 @@ def _boolize(data):
 def validate_filename():
     """Validate filename for invalid characters/parts."""
     return jsonify(
-        filename=_validate_filename(request.form['filename'])
+        filename=do_validate_filename(request.form['filename'])
     )
-
-
-def _validate_filename(filename):
-    illegalords = range(0, 32) + [127]
-    for char in filename:
-        assert char not in '[]{}|#<>%+?:/\\', \
-            'Your filename contains an illegal character: ' + char
-
-        # ord(char) to prevent bad renderings
-        assert ord(char) not in illegalords, \
-            'Your filename contains an illegal character: chr(%d)' % ord(char)
-
-    assert len(filename) < 250, 'Your filename is too long'
-
-    assert not re.search(r"&[A-Za-z0-9\x80-\xff]+;", filename), \
-        'Your filename contains XML/HTML character references'
-
-    return filename.replace('_', ' ')
 
 
 def get_backend_keys(format):

--- a/video2commons/frontend/api.py
+++ b/video2commons/frontend/api.py
@@ -68,7 +68,7 @@ def format_exception(e):
     if isinstance(e, AssertionError):
         return desc
     else:
-        return 'An exception occured: %s: %s' % \
+        return 'An exception occurred: %s: %s' % \
             (type(e).__name__, desc)
 
 

--- a/video2commons/frontend/urlextract.py
+++ b/video2commons/frontend/urlextract.py
@@ -245,7 +245,7 @@ def do_validate_filename(filename):
 
     for rule in sanitationRules:
         reobj = rule['pattern'].search(filename)
-        assert not reobj or reobj.group(0) != ' ', \
+        assert not reobj or reobj.group(0) == ' ', \
             'Your filename contains an illegal part: %r' % reobj.group(0)
 
     return filename.replace('_', ' ')

--- a/video2commons/frontend/urlextract.py
+++ b/video2commons/frontend/urlextract.py
@@ -225,7 +225,7 @@ sanitationRules = [
     },
     # everything that wasn't covered yet
     {
-        'pattern': re.compile(ur'[|#%+?:/\\\u0000-\u001f\u007f]'),
+        'pattern': re.compile(ur'[|#+?:/\\\u0000-\u001f\u007f]'),
         'replace': '-'
     },
 ]

--- a/video2commons/frontend/urlextract.py
+++ b/video2commons/frontend/urlextract.py
@@ -193,13 +193,13 @@ sanitationRules = [
     # URL encoding (possibly)
     {
         'pattern': re.compile(ur'%([0-9A-Fa-f]{2})'),
-        'replace': '% $1'
+        'replace': r'% \1'
     },
     # HTML-character-entities
     {
         'pattern': re.compile(ur'&(([A-Za-z0-9\x80-\xff]+|'
                               ur'#[0-9]+|#x[0-9A-Fa-f]+);)'),
-        'replace': '& $1'
+        'replace': r'& \1'
     },
     # slash, colon (not supported by file systems like NTFS/Windows,
     # Mac OS 9 [:], ext4 [/])
@@ -245,7 +245,7 @@ def do_validate_filename(filename):
 
     for rule in sanitationRules:
         reobj = rule['pattern'].search(filename)
-        assert not reobj, \
+        assert not reobj or reobj.group(0) != ' ', \
             'Your filename contains an illegal part: %r' % reobj.group(0)
 
     return filename.replace('_', ' ')

--- a/video2commons/frontend/urlextract.py
+++ b/video2commons/frontend/urlextract.py
@@ -74,7 +74,7 @@ def do_extract_url(url):
         'url': url,
         'extractor': ie_key,
         'filedesc': filedesc.strip(),
-        'filename': title
+        'filename': sanitize(title)
     }
 
 
@@ -165,3 +165,87 @@ def escape_wikitext(wikitext):
     rep = dict((re.escape(k), v) for k, v in rep.iteritems())
     pattern = re.compile("|".join(rep.keys()))
     return pattern.sub(lambda m: rep[re.escape(m.group(0))], wikitext)
+
+
+# Source: mediawiki.Title.js@9df363d
+sanitationRules = [
+    # "signature"
+    {
+        'pattern': re.compile(ur'~{3}'),
+        'replace': ''
+    },
+    # Space, underscore, tab, NBSP and other unusual spaces
+    {
+        'pattern': re.compile(ur'[ _\u0009\u00A0\u1680\u180E\u2000-\u200A'
+                              ur'\u2028\u2029\u202F\u205F\u3000\s]+'),
+        'replace': ' '
+    },
+    # unicode bidi override characters: Implicit, Embeds, Overrides
+    {
+        'pattern': re.compile(ur'[\u200E\u200F\u202A-\u202E]'),
+        'replace': ''
+    },
+    # control characters
+    {
+        'pattern': re.compile(ur'[\x00-\x1f\x7f]'),
+        'replace': ''
+    },
+    # URL encoding (possibly)
+    {
+        'pattern': re.compile(ur'%([0-9A-Fa-f]{2})'),
+        'replace': '% $1'
+    },
+    # HTML-character-entities
+    {
+        'pattern': re.compile(ur'&(([A-Za-z0-9\x80-\xff]+|'
+                              ur'#[0-9]+|#x[0-9A-Fa-f]+);)'),
+        'replace': '& $1'
+    },
+    # slash, colon (not supported by file systems like NTFS/Windows,
+    # Mac OS 9 [:], ext4 [/])
+    {
+        'pattern': re.compile(ur'[:/#]'),
+        'replace': '-'
+    },
+    # brackets, greater than
+    {
+        'pattern': re.compile(ur'[\]\}>]'),
+        'replace': ')'
+    },
+    # brackets, lower than
+    {
+        'pattern': re.compile(ur'[\[\{<]'),
+        'replace': '('
+    },
+    # directory structures
+    {
+        'pattern': re.compile(ur'^(\.|\.\.|\./.*|\.\./.*|.*/\./.*|'
+                              ur'.*/\.\./.*|.*/\.|.*/\.\.)$'),
+        'replace': ''
+    },
+    # everything that wasn't covered yet
+    {
+        'pattern': re.compile(ur'[|#%+?:/\\\u0000-\u001f\u007f]'),
+        'replace': '-'
+    },
+]
+
+
+def sanitize(filename):
+    """Sanitize a filename for uploading."""
+    for rule in sanitationRules:
+        filename = rule['pattern'].sub(rule['replace'], filename)
+
+    return filename
+
+
+def do_validate_filename(filename):
+    """Validate filename for invalid characters/parts."""
+    assert len(filename) < 250, 'Your filename is too long'
+
+    for rule in sanitationRules:
+        reobj = rule['pattern'].search(filename)
+        assert not reobj, \
+            'Your filename contains an illegal part: %r' % reobj.group(0)
+
+    return filename.replace('_', ' ')

